File: CS/EditManyToMany/XpoHelper.cs
using System;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.Metadata;

/// <summary>
/// Summary description for XpoHelper
/// </summary>
public static class XpoHelper {
    static XpoHelper() {
        UpdateSchema();
        CreateDefaultObjects();
    }

    public static Session GetNewSession() {
        return new Session(DataLayer);
    }

    public static UnitOfWork GetNewUnitOfWork() {
        return new UnitOfWork(DataLayer);
    }

    private readonly static object lockObject = new object();

    static IDataLayer fDataLayer;
    static IDataLayer DataLayer {
        get {
            if(fDataLayer == null) {
                lock(lockObject) {
                    fDataLayer = GetDataLayer();
                }
            }
            return fDataLayer;
        }
    }

    private static IDataLayer GetDataLayer() {
        XpoDefault.Session = null;

        InMemoryDataStore provider = new InMemoryDataStore();
        IDataLayer dl = new SimpleDataLayer(provider);

        return dl;
    }

    private static void UpdateSchema() {
        GetNewSession().UpdateSchema(typeof(DataObjects.Department).Assembly);
    }

    private static void CreateDefaultObjects() {
        using(UnitOfWork uow = GetNewUnitOfWork()) {
            DataObjects.Department dept = new DataObjects.Department(uow);
            dept.Name = "Dept A";
            DataObjects.Location loc = new DataObjects.Location(uow);
            loc.Region = "North";
            loc.Departments.Add(dept);
            uow.CommitChanges();
        }
    }
}

File: VB/EditManyToMany/XpoHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.Metadata

''' <summary>
''' Summary description for XpoHelper
''' </summary>
Public NotInheritable Class XpoHelper
	Private Sub New()
	End Sub
	Shared Sub New()
		UpdateSchema()
		CreateDefaultObjects()
	End Sub

	Public Shared Function GetNewSession() As Session
		Return New Session(DataLayer)
	End Function

	Public Shared Function GetNewUnitOfWork() As UnitOfWork
		Return New UnitOfWork(DataLayer)
	End Function

	Private ReadOnly Shared lockObject As Object = New Object()

	Private Shared fDataLayer As IDataLayer
	Private Shared ReadOnly Property DataLayer() As IDataLayer
		Get
			If fDataLayer Is Nothing Then
				SyncLock lockObject
					fDataLayer = GetDataLayer()
				End SyncLock
			End If
			Return fDataLayer
		End Get
	End Property

	Private Shared Function GetDataLayer() As IDataLayer
		XpoDefault.Session = Nothing

		Dim provider As New InMemoryDataStore()
		Dim dl As IDataLayer = New SimpleDataLayer(provider)

		Return dl
	End Function

	Private Shared Sub UpdateSchema()
		GetNewSession().UpdateSchema(GetType(DataObjects.Department).Assembly)
	End Sub

	Private Shared Sub CreateDefaultObjects()
		Using uow As UnitOfWork = GetNewUnitOfWork()
			Dim dept As New DataObjects.Department(uow)
			dept.Name = "Dept A"
			Dim loc As New DataObjects.Location(uow)
			loc.Region = "North"
			loc.Departments.Add(dept)
			uow.CommitChanges()
		End Using
	End Sub
End Class
